Add a webkitgtk_minibrowser product/browser.

clopez opened the pull request:
* The current "webkit" product is a bit of confusing because
   it is a too much generic name, and it can be used for
   running webkitgtk or wpewebkit based browsers (but not for
   safari/mac based ones)

 * This adds a webkitgtk_minibrowser product that makes much
   clearer what is the product about. It also defines default
   values for all variables that previously the user had to
   fill when running the "webkit" product. Now it simply
   suffices with running: "/wpt run webkitgtk_minibrowser"

clopez committed:
Add a webkitgtk_minibrowser product/browser.

 * The current "webkit" product is a bit of confusing because
   it is a too much generic name, and it can be used for
   running webkitgtk or wpewebkit based browsers (but not for
   safari/mac based ones)

 * This adds a webkitgtk_minibrowser product that makes much
   clearer what is the product about. It also defines default
   values for all variables that previously the user had to
   fill when running the "webkit" product. Now it simply
   suffices with running: "/wpt run webkitgtk_minibrowser"

clopez commented:
@foolip @zdobersek This patch doesn't remove the product "webkit", but I'm considering doing that  and adding a third "wpewebkit_minibrowser" to cover the last use case of this "webkit" product not handled by the other products/browsers.

So (if I do that) we would end with 3 products/browsers based on webkit/linux: epiphany, webkitgtk_minibrowser and wpewebkit_minibrowser.

That way instead of running `./wpt run --webkit-port wpe webkit` you would run `./wpt run wpewebkit_minibrowser`

Let me know if you think this is a good/bad idea.

At least from the point of view of the json files generated I find much clearer to see product "webkitgtk_minibrowser" than product "webkit"

foolip commented on tools/wptrunner/wptrunner/browsers/webkitgtk_minibrowser.py:
```diff
@@ -0,0 +1,74 @@
+
+    return {
+        "browserName": "MiniBrowser",
+        "browserVersion": "2.20",
```
Is this the payload for the POST /session request? If so is the version required, given that the pass the binary it seems like nothing useful could be done with it?

clopez replied:
Not sure if its required. I copied it from the current "webkit" version. i will check if its needed.

clopez replied:
It seems `browserVersion` acts as the minimum version that would be allowed. So with this code, the webdriver would refuse to start the MiniBrowser if it reports to be a version lower than < 2.20 runninganything newer than "2.20". The f[irst release of webkitgtk to support webdriver was 2.18](https://blogs.igalia.com/carlosgc/2017/09/09/webdriver-support-in-webkitgtk-2-18/). So I guess it makes sense to lower this to 2.18

foolip replied:
Isn't it allowed to omit it entirely? I known that safaridriver at least is able to start a session with an empty capabilities dictionary. Not sure what the spec requires.

clopez replied:
Yes, it is allowed to omit the version entirely. If its omitted, then it will start the browser without checking the version.

foolip replied:
That sounds better, right?

clopez replied:
Right :)

foolip commented on tools/wptrunner/wptrunner/browsers/webkitgtk_minibrowser.py:
```diff
@@ -0,0 +1,74 @@
+        "webkitgtk:browserOptions": {
+            "binary": kwargs["binary"],
+            "args": args,
+            "certificates": [
```
Is this still needed? If it worked I presume https://github.com/web-platform-tests/wpt/pull/19185 wouldn't be needed?

clopez replied:
Good point. I guess this was also copied from other browser drivers, but it doesn't work for WebKitWebDriver.

clopez replied:
Mmmm. Looking at the WebKit code it seems it should support it. I will try to check what is wrong here.

clopez replied:
Ok. This seems to be working as expected. Running the test `content-security-policy/worker-src/service-list.https.sub.html`on my dev machine works both with Epiphany and the WebKitGTK MiniBrowser (no cert error) and the url loads via https. And I don't have that cert installed globaly as opening that very same url in another epiphany window (new process) gives an https certificate error.

However, running the test that appears on the screenshot at https://github.com/web-platform-tests/wpt/pull/18595#issuecomment-531163125 (/cookies/samesite/fetch.https.html) gives me a TLS error if the cert is not installed globally. 

I have hunted this down to be a problem on how wpt tells webkitwebdriver for which domains the certificate is valid. I reported this on #19233

foolip commented on tools/wptrunner/wptrunner/browsers/webkitgtk_minibrowser.py:
```diff
@@ -0,0 +1,74 @@
+    return {
+        "browserName": "MiniBrowser",
+        "browserVersion": "2.20",
+        "platformName": "ANY",
```
Is `platformName` required?

clopez replied:
`platforName` acts as a filter. The webdriver would refuse to start the browser if the platform it reports (would be: linux) doesn't match was asked. "any" acts as a wildcard. I guess I can remove this as that would have the same effect than specifying "any" here.

foolip replied:
Gone in https://github.com/web-platform-tests/wpt/pull/19227/commits/462c2dfbf7534d944af4fda1081018f8f6b21f0c.

foolip commented on tools/wpt/browser.py:
```diff
@@ -1104,6 +1104,38 @@ def version(self, binary=None, webdriver_binary=None):
+        try: # WebKitGTK MiniBrowser before 2.26.0 doesn't support --version
+            output = call(binary, "--version").strip()
+        except subprocess.CalledProcessError:
+            output = None
```
```suggestion
            return None
```

foolip replied:
Ping @clopez on this suggestion.

foolip replied:
Oh actually you included the change in https://github.com/web-platform-tests/wpt/pull/19227/commits/462c2dfbf7534d944af4fda1081018f8f6b21f0c.

foolip commented on tools/wpt/browser.py:
```diff
@@ -1104,6 +1104,38 @@ def version(self, binary=None, webdriver_binary=None):
+        except subprocess.CalledProcessError:
+            output = None
+        if output:
+            return output.split()[1]
```
What is the first component? Can you comment on the output format and add a test similar to here:?
https://github.com/web-platform-tests/wpt/blob/master/tools/wpt/browser.py#L983
https://github.com/web-platform-tests/wpt/blob/master/tools/wpt/tests/test_browser.py#L12

clopez replied:
The output looks like this:
```
WebKitGTK 2.26.1
```
I will add a test.

foolip replied:
Matching a regexp like in `r"Included with Safari (.*)"` would give some assurance that if the output format changes this stops matching instead of returning some other part of the output that isn't a version numbers.

foolip replied:
Fixed in https://github.com/web-platform-tests/wpt/pull/19227/commits/462c2dfbf7534d944af4fda1081018f8f6b21f0c.

foolip commented on tools/wpt/browser.py:
```diff
@@ -1104,6 +1104,38 @@ def version(self, binary=None, webdriver_binary=None):
+class WebKitGTKMiniBrowser(WebKit):
+
+    def find_binary(self, venv_path=None, channel=None):
+        libexecpaths = ["/usr/libexec/webkit2gtk-4.0",
```
How often does the "4.0" part of this path change, and what will happen when it does?

foolip replied:
Alright, sounds like a risk we can accept :)

foolip commented on tools/wpt/browser.py:
```diff
@@ -1104,6 +1104,38 @@ def version(self, binary=None, webdriver_binary=None):
+        return find_executable("MiniBrowser", os.pathsep.join(libexecpaths))
+
+    def find_webdriver(self, channel=None):
+        return find_executable("WebKitWebDriver")
```
Note to other reviewers: searching packages.debian.org suggests this is in the webkit2gtk-driver package, that's why the lookup has nothing special about it.

foolip commented on tools/wpt/browser.py:
```diff
@@ -1104,6 +1104,38 @@ def version(self, binary=None, webdriver_binary=None):
+    def find_binary(self, venv_path=None, channel=None):
+        libexecpaths = ["/usr/libexec/webkit2gtk-4.0",
+                        "/usr/lib/x86_64-linux-gnu/webkit2gtk-4.0"]
+        # If GCC is available use it to detect this machine triplet,
```
This is a bit convoluted, and I expect it's hard to guess what one has to install to run this product, or to work backwards from this to what the Debian package name is.

Can you add a piece of documentation like https://github.com/web-platform-tests/wpt/blob/master/docs/running-tests/chrome.md that says what to packages to install to run this product?

clopez replied:
Sure

foolip commented on tools/wpt/run.py:
```diff
@@ -583,6 +583,29 @@ def setup_kwargs(self, kwargs):
+            binary = self.browser.find_binary()
+
+            if binary is None:
+                raise WptrunError("Unable to find MiniBrowser binary")
```
I notice this doesn't say "in PATH". If it it in PATH, will it be found there?

clopez replied:
current method doesn't try to find MiniBrowser binary in PATH because its usually installed inside the webkit2gtk-4.0 directory of libexecdir and not in PATH.

foolip reviewed (commented):
Adding a new product makes sense to me, and the suggested name is good. Some nits.

clopez committed:
webkitgtk_minibrowser: add documentation, tests and simplify capabilities.

* The capabilities requested to the WebKitWebDriver don't need to specify
platformName and browserVersion. If those are not specified the driver will
start the requested browser without checking this values. Simplify it.

* Add documentation about what packages are needed for Fedora/Debian/Ubuntu
to test with the WebKitGTK MiniBrowser and how to launch it.

* Add unit tests to check webkitgtk_minibrowser version() and find_binary()
functions.

foolip commented on docs/running-tests/webkitgtk_minibrowser.md:
```diff
@@ -0,0 +1,32 @@
+# WebKitGTK MiniBrowser
```
Good stuff!

guest271314 replied:
@foolip Does this PR include Epiphany Technology Preview within WebKitGTK MiniBrowser?

clopez replied:
No, this PR is unrelated to Epiphany.
For running tests with Epiphany, you need to use the browser "epiphany" for "wpt run"

foolip commented:
@clopez just so that there's clarity on who's waiting for whom, next steps are for you to push some changes to simplify, right? I think what's remaining is mostly nits and some tests.

clopez commented:
> @clopez just so that there's clarity on who's waiting for whom, next steps are for you to push some changes to simplify, right? I think what's remaining is mostly nits and some tests.

I think I already did that on the second commit on this PR yesterday.
This second commit included not only documentation, but also simplification of the capabilities negotiation and pytest unit tests.

What is exactly still missing? Thanks

foolip commented:
Oh, the reason I didn't think it was done is because there are still unresolved review threads that end with me asking a question. Can you resolve all the conversations if they're done? That makes the state of things much clearer.

clopez commented:
Oh, I see. I didn't knew about this "resolve conversation" thing. I have done that now :)

foolip commented:
The certificates setup is still odd, but that's tracked by https://github.com/web-platform-tests/wpt/issues/19233 now.

foolip reviewed (approved):
Excellent tests!

foolip commented:
Alright, I think all that remains now is fixing failing in CI, both Taskcluster and Azure Pipelines. Note that https://github.com/web-platform-tests/wpt/issues/19297 might be behind some of the failures, sorry if that makes it hard to navigate the CI failures.

clopez committed:
fixup! webkitgtk_minibrowser: add documentation, tests and simplify capabilities.

clopez committed:
fixup! webkitgtk_minibrowser: add documentation, tests and simplify capabilities.

clopez committed:
fixup! webkitgtk_minibrowser: add documentation, tests and simplify capabilities.

clopez commented:
@foolip Now all tests are passing on Azure. On taskcluster all are passing except `infrastructure/ tests`. But I think [the failure](https://tools.taskcluster.net/groups/TdxfGKJnSeeNrFqae0aeMQ/tasks/beMiddthQuCsQhpCLHRLgw/runs/0/logs/public%2Flogs%2Flive.log) is unrelated with the changes on this PR.

foolip commented:
Yeah, I think that failure is caused by the Chrome Dev update. I'll admin merge this.
